### PR TITLE
General wall distance functions

### PR DESCRIPTION
**Pull request by AdelekeBankole**

### CODE REVIEW

File: examples/fluids/navierstokes.h [hunk 1]
```diff
@@ -140,6 +140,14 @@ struct CeedData_private {
       qf_apply_outflow_jacobian, qf_apply_freestream, qf_apply_freestream_jacobian;
 };
 
+// Distance functions
+struct Distance_private {
+  DM           dm;
+  KSP          kspDist;
+  SNES         snesDist;
```
Comment (jedbrown) [inline examples/fluids/navierstokes.h:147]: One should be enough.

File: examples/fluids/src/setupdm.c [hunk 2]
```diff
@@ -71,6 +71,11 @@ PetscErrorCode SetUpDM(DM dm, ProblemData *problem, PetscInt degree, SimpleBC bc
       PetscCall(DMAddBoundary(dm, DM_BC_ESSENTIAL, "slipz", label, bc->num_slip[2], bc->slips[2], 0, 1, comps, (void (*)(void))NULL, NULL,
                               problem->bc_ctx, NULL));
     }
+    // Set dirichlet boundary condition on cylinder walls
+    if (bc->num_wall > 0) {
+      PetscCall(DMAddBoundary(dm, DM_BC_ESSENTIAL, "cylinder_wall", label, bc->num_wall, bc->walls, 0, 5, bc->wall_comps, (void (*)(void))problem->bc,
+                              NULL, problem->bc_ctx, NULL));
+    }
```
Comment (jedbrown) [inline examples/fluids/src/setupdm.c:78]: I think this should be setting up the distance DM, which is a scalar field (thus only one component).

File: examples/fluids/navierstokes.h (hunk 1 above)
Comment (AdelekeBankole) [reply examples/fluids/navierstokes.h:147]: Alright, ok

File: examples/fluids/src/setupdm.c [hunk 3]
```diff
@@ -71,6 +71,11 @@ PetscErrorCode SetUpDM(DM dm, ProblemData *problem, PetscInt degree, SimpleBC bc
       PetscCall(DMAddBoundary(dm, DM_BC_ESSENTIAL, "slipz", label, bc->num_slip[2], bc->slips[2], 0, 1, comps, (void (*)(void))NULL, NULL,
                               problem->bc_ctx, NULL));
     }
+    // Set dirichlet boundary condition on cylinder walls
+    if (bc->num_wall > 0) {
+      PetscCall(DMAddBoundary(dm, DM_BC_ESSENTIAL, "cylinder_wall", label, bc->num_wall, bc->walls, 0, 1, bc->wall_comps, (void (*)(void))problem->bc,
+                              NULL, problem->bc_ctx, NULL));
+    }
```
Comment (jedbrown) [inline examples/fluids/src/setupdm.c:78]: These BCs need to be set on a new DM for solving the distance equation. That DM will have only one component.

File: examples/fluids/src/setupts.c [hunk 4]
```diff
@@ -179,6 +179,29 @@ static PetscErrorCode Surface_Forces_NS(DM dm, Vec G_loc, PetscInt num_walls, co
...
+  MPI_Comm comm;
+
+  PetscFunctionBeginUser;
+  PetscBool distance_snes_monitor = PETSC_FALSE;
+  PetscCall(PetscOptionsHasName(NULL, NULL, "-distance_snes_monitor", &distance_snes_monitor));
+  PetscCall(PetscObjectGetComm((PetscObject)dm, &comm));
+  PetscCall(SNESCreate(PETSC_COMM_WORLD, &snesDist));
+  PetscObjectSetOptionsPrefix((PetscObject)snesDist, "distance_");
+  PetscCall(PetscMalloc1(dim * num_wall, &Dist_loc));
+  PetscCall(DMClone(dm, &dmDist));
```
Comment (jedbrown) [inline examples/fluids/src/setupts.c:198]: You'll need to add a field here, with one component.

**Review by jedbrown — COMMENTED** [review]
- [ ] Create a CeedQFunction to evaluate $\nabla v \cdot \nabla u - v 1$.
- [ ] Create a CeedOperator for `dmDist`
- [ ] Solve using `-snes_fd_color` (won't be efficient)
- [ ] Port multigrid from `examples/petsc/multigrid.c` to make it efficient
- [ ] Given solution $u$, compute distance $d = \sqrt{\lVert \nabla \phi \rVert^2 + 2\phi} - \lVert \nabla \phi \rVert$.
- [ ] Visualize $d$ field and compare with analytic distance

File: examples/fluids/src/setupts.c [hunk 5]
```diff
@@ -179,6 +179,29 @@ static PetscErrorCode Surface_Forces_NS(DM dm, Vec G_loc, PetscInt num_walls, co
...
+  SNES     snesDist;
+  MPI_Comm comm;
+
+  PetscFunctionBeginUser;
+  PetscBool distance_snes_monitor = PETSC_FALSE;
+  PetscCall(PetscOptionsHasName(NULL, NULL, "-distance_snes_monitor", &distance_snes_monitor));
+  PetscCall(PetscObjectGetComm((PetscObject)dm, &comm));
+  PetscCall(SNESCreate(PETSC_COMM_WORLD, &snesDist));
+  PetscObjectSetOptionsPrefix((PetscObject)snesDist, "distance_");
+  PetscCall(PetscMalloc1(dim * num_wall, &Dist_loc));
```
Comment (jedbrown) [inline examples/fluids/src/setupts.c:197]: We don't need this.

**Comment by AdelekeBankole** [comment]
@jedbrown, I am doing the substraction in here ```CeedQFunctionCreateInterior(ceed, dim, DistanceFunction - Mass_N, DistanceFunction_loc, &qf_distance_function);```  because this is what I want to do $F(u) = Au - M 1$
Does this make sense?

**Comment by jeremylt** [comment]
Closing in anticipation of moving Honee to a new repo

**Pull request closed without merging**